### PR TITLE
Giving better client-side feedback on websocket authT/authZ issues

### PR DESCRIPTION
**Pull request by chrisguidry**
In Prefect Cloud, we're now giving better error messages when authentication
fails on a websocket connection, and this change has the clients show those
error reasons to the users in the exceptions thrown during
connection/reconnection.

### CODE REVIEW

File: src/prefect/events/clients.py [hunk 1]
```diff
@@ -330,14 +330,20 @@ async def _reconnect(self) -> None:
         )
 
         try:
-            message = orjson.loads(await self._websocket.recv())
+            message: dict[str, Any] = orjson.loads(await self._websocket.recv())
```
Comment (zzstoatzz) [inline src/prefect/events/clients.py:333]: i think we need `Dict` here right? (for 3.8)

```python
(oss-3.8) pad-2 :: github.com/PrefectHQ/prefect
» python --version
Python 3.8.17

» python -c "x: dict[str, str] = {}"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: 'type' object is not subscriptable
```

File: src/prefect/client/subscriptions.py [hunk 2]
```diff
@@ -71,8 +71,8 @@ async def _ensure_connected(self):
                 ).decode()
             )
 
-            auth = orjson.loads(await websocket.recv())
-            assert auth["type"] == "auth_success"
+            auth: dict[str, Any] = orjson.loads(await websocket.recv())
```
Comment (zzstoatzz) [inline src/prefect/client/subscriptions.py:74]: same as other comment

File: src/prefect/events/clients.py (hunk 1 above)
Comment (zzstoatzz) [reply src/prefect/events/clients.py:333]: looks like we might have also let this slip into `tests/events/test_v1_triggers_deserialization.py`

Comment (chrisguidry) [reply src/prefect/events/clients.py:333]: Yikes, thank you!  I think we lost our ability to catch these in type-checking somewhere along the way.  I'll try to get that back.

Comment (chrisguidry) [reply src/prefect/events/clients.py:333]: This is really weird, the tests aren't failing in 3.8 :thinking:

Comment (zzstoatzz) [reply src/prefect/events/clients.py:333]: hmm that is weird 🧐

**Review by zzstoatzz — APPROVED** [review]

File: src/prefect/events/clients.py (hunk 1 above)
Comment (bunchesofdonald) [reply src/prefect/events/clients.py]: Perhaps relatedly, I was messing around with `vermin` (https://github.com/netromdk/vermin) and it found several issues in the codebase for 3.8:

```
src/prefect/client/subscriptions.py:77::!2:3.9:dict union (dict | dict)
src/prefect/events/related.py:77::!2:3.9:builtin generic type annotation (list[..])
src/prefect/server/database/configurations.py:33::!2:3.9:builtin generic type annotation (contextvars.ContextVar[..])
src/prefect/task_server.py:73::!2:3.9:builtin generic type annotation (list[..])
```

**Comment by chrisguidry** [comment]
Yes!  I was just thinking about this PR when I saw your other one @bunchesofdonald .  I'd say we should definitely get `vermin` into our pre-commit here in OSS.

**Pull request merged**